faux-mgs/src/usart.rs

```diff
@@ -138,29 +163,87 @@ pub(crate) async fn run(
             }
         }
     }
+
+    // Drop the sending half of this channel to signal our tx-to-sp task to
+    // exit.
+    mem::drop(send_tx);
+    let console_tx = tx_to_sp_handle.await.expect("tx_to_sp task panicked");
+
+    // If we encountered a fatal error, we will not attempt to detach from the
+    // SP. One possible fatal error is _someone else detached us_, and they
+    // might have attached themselves in the meantime. Other fatal errors
+    // indicate a serious problem communicating with the SP, and it's likely
+    // detaching will fail anyway.
+    if !encountered_fatal_error {
+        console_tx
+            .detach()
+            .await
+            .context("failed to detach from SP console")?;
+    }
+
+    // Restore termios settings, if we put the terminal into raw mode. (This
+    // would happen automatically when the guard is dropped, but doing it
+    // explicitly lets us check for errors.)
+    if let Some(guard) = termios_guard {
+        guard.restore()?;
+    }
+
+    Ok(())
 }
 
 async fn relay_data_to_sp(
     mut console_tx: AttachedSerialConsoleSend,
     mut data_rx: mpsc::Receiver<SendTxData>,
+    fatal_err_tx: oneshot::Sender<CommunicationError>,
     log: slog::Logger,
-) -> Result<()> {
+) -> AttachedSerialConsoleSend {
     let mut keepalive = time::interval(SERIAL_CONSOLE_IDLE_TIMEOUT / 4);
     keepalive.set_missed_tick_behavior(MissedTickBehavior::Delay);
 
+    // If we fail to send a message, we need to resend it; we'll keep a running
+    // ticker and check for messages-to-send every second.
+    let mut check_for_resend = time::interval(Duration::from_secs(1));
+    check_for_resend.set_missed_tick_behavior(MissedTickBehavior::Skip);
+
+    let mut messages_to_send = VecDeque::new();
```

Comment:
Presumably, this is only one or two messages long right?

Reply:
Got it. Still, human driven, and self imposed memory bloat based on text input. Seems fine!